rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,227 @@
+on the backend, and then will decide on appropriate actions (create/update/delete) that are required on
+the backend to change resources to the desired state.
+The Teleport app service that runs in the same Kubernetes cluster will then react to those changes, and will start/stop proxying
+apps that are selected by appropriate labels.
+
+Name of the created Teleport app will consist of Kubernetes service name, namespace and cluster name:
+`$SERVICE_NAME-$CLUSTER_NAME-$NAMESPACE`. This is done to avoid naming collisions in discovered apps. Though app names
+will be long in that case, there's ongoing work to improve UX when working with long resources names in Teleport, see [RFD 129](https://github.com/gravitational/teleport/pull/27258).
+
+We will add capability to use optional annotations to better control transformation of Kubernetes services into Teleport apps.
+By default we will be exposing services as `tcp` apps, since it's the most general type, but annotation or exposed port's `appProtocol` field
+can be used to specify which protocol to use for the app's URI. 
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels:
+        env: staging
+    - types: ["app"]
```

Comment:
IMO "types: app" is a little confusing. In other discovery configurations, the "types" field represents the thing that is being discovered (rds, eks, ec2, etc.), but here it represents what the discovered thing is being translated into in Teleport. So something like "types: service" would probably be more consistent, although I understand that the service can represent both an app and (in future) a database.

TBH, I would probably just remove the "types" field for now, until we have a need for it. WDYT?

Reply:
I think they are close enough in meaning, don't really think it will be confusing. It's more of "looking for apps among services" and "looking for DBs among service", because they are all services by default in k8s space.
Yep, we technically don't need this field right now. I would leave it, since we know we want to database, but I'm not tied to it.

Reply:
The initial idea for introducing this config was to identify the type of objects that the labels should create so we don't have to find out - some dbs support HTTP which would be problematic.

We can remove it but when adding other types like databases it will be difficult to maintain compatibility without breaking the initial configuration

Reply:
Alright, it's probably ok even if a little inconsistent with other matchers. I was also thinking what if we want to add support, for instance, for discovering [Application CRDs](https://github.com/kubernetes-sigs/application) in the cluster in future, but we can just extend the `app` type to support search for both services and application custom resources.

Reply:
Just a heads up about the application CRD:

> An Application object provides a way for you to aggregate individual Kubernetes components (e.g. Services, Deployments, StatefulSets, Ingresses, CRDs), and manage them as a group. 

Services & Deployments still exist in a world where people use Application CRDs so it won't affect our discovery process. App CRD just aggregates and gives context to a group of Kube resources